[IMP] base_import: support multiple xlsx sheets

Julien00859 opened the pull request:
Excel can store data in multiple tables, each is called a sheet. Odoo
was only importing the first sheet making the process to import a file
containing multiple sheets cumbersome.

It is now possible to select the sheet in the import options.

Task: 2043768

xmo-odoo commented on addons/base_import/static/src/js/import_action.js:
```diff
@@ -400,8 +421,16 @@ var DataImport = AbstractAction.extend({
+        // if sheet names changed, equivalent for python's set(a) != set(b)
+        if ((this.xlsx_sheets.length != result.options.sheets.length)
+            || (_.intersection(this.xlsx_sheets, result.options.sheets).length
+                != _.uniq(this.xlsx_sheets.length))) {
```
Why not just use _.isEqual?

Julien00859 replied:
because I didn't know it existed

xmo-odoo commented on addons/base_import/static/src/xml/base_import.xml:
```diff
@@ -35,6 +35,22 @@
                         </div>
                 </div>
+                <div class="oe_import_with_file oe_import_xlsx row">
+                    <a href="#" class="oe_import_xlsx_toggle col-sm-12">
```
The regular import options are toggled because there's several and they're quite complex. Here given for now there's just the sheet picker, the toggle should probably be removed. *However* the box should be hidden if there's only one sheet, no need to povide a choice between one option.

Julien00859 replied:
ngr thinks the same, I'm moving the option elsewhere

xmo-odoo commented:
Also rather than the old check on getting the result of filetype guessing, the choice between CSV and spreadsheet could be done by just checking if the loaded file returns a `sheets` option: CSV doesn't, the rest should.

Julien00859 commented:
HTML
* removed the option menu in favor of just an input

JS
* `xlsx_options` => `spreadsheet_opts`
* `this.xlsx_sheets` => `this.spreadsheets`
* restored the CSV menu triggering code
* added a toggle to show/hide the select

Python
* `sheet` => `spreadsheet`
* `sheets` => `spreadsheets`

xmo-odoo commented:
> this.xlsx_sheets => this.spreadsheets
> 
> sheet => spreadsheet
> sheets => spreadsheets

That should not have been modified: a spreadsheet (file) contains 1..n (work)sheets.

Julien00859 commented:
* rollbacked sheet and sheets
* set a select2 option to enable the searchbar at 10+ elements

xmo-odoo commented on addons/base_import/static/src/js/import_action.js:
```diff
@@ -146,6 +149,7 @@ var DataImport = AbstractAction.extend({
         this.session = session;
         this._title = _t('Import a File'); // Displayed in the breadcrumbs
         this.do_not_change_match = false;
+        this.sheets = [];
```
This should probably be set to null, not an array. Sheet data is per-load information (it can change when reloading a file or loading a new file), things don't work properly otherwise.

xmo-odoo commented on addons/base_import/static/src/js/import_action.js:
```diff
@@ -168,6 +172,7 @@ var DataImport = AbstractAction.extend({
         this.setup_separator_picker();
         this.setup_float_format_picker();
         this.setup_date_format_picker();
+        this.setup_sheets_picker();
```
This may be a bad idea? The sheet picker *needs* remote data to be properly set up. Other pickers set up here have completely static inputs. I'm not sure select2 can just be re-invoked / re-setup on an input either. It probably needs to either be destroyed and recreated, or needs to be *updated* rather than reinvoked.

xmo-odoo commented:
Still missing sheets/sheet support in ODS import.

xmo-odoo commented:
@robodoo r+
